src/helpers.rs

```diff
@@ -344,4 +345,57 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
         Ok(())
     }
+
+    fn read_os_string_from_c_string(&mut self, scalar: Scalar<Tag>) -> InterpResult<'tcx, OsString> {
+        let bytes = self.eval_context_mut().memory.read_c_str(scalar)?;
+        Ok(bytes_to_os_str(bytes)?.into())
+    }
+
+    fn write_os_str_to_c_string(&mut self, os_str: &OsStr, ptr: Pointer<Tag>, size: u64) -> InterpResult<'tcx> {
+        let bytes = os_str_to_bytes(os_str)?;
+        let len = bytes.len();
+        // If `size` is smaller or equal than `bytes.len()`, writing `bytes` plus the required null
+        // terminator to memory using the `ptr` pointer would cause an overflow.
+        if (len as u64) < size {
+            let this = self.eval_context_mut();
+            let tcx = &{ this.tcx.tcx };
+            let buffer = this.memory.get_mut(ptr.alloc_id)?.get_bytes_mut(tcx, ptr, Size::from_bytes(len as u64 + 1))?;
```

Comment:
This addition might overflow, then everything will go horribly wrong.

Reply:
Also this doesn't do the bounds check. Please read the doc-comments of the functions you are calling! For [`get_bytes_mut`, it says](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/interpret/struct.Allocation.html#method.get_bytes_mut)

> It is the caller's responsibility to check bounds and alignment beforehand.

I think the better approach here is to add, on the rustc side, a `write_bytes` method mirroring the `read_bytes` that we already have. Then you can use that here. If you implement that not for `&[u8]` but for `impl Iterator<Item=u8> + TrustedLen` you can even handle adding the trailing 0 by using `.chain`. :D

Reply:
Can we do some workaround first and then do that one? this PR is getting old :P

Reply:
> This addition might overflow, then everything will go horribly wrong.

Should I check for the overflow and throw_unsup then?

Reply:
See: https://github.com/rust-lang/miri/pull/993/commits/f7c6e0efbe11343562c713dce1d28e2d141dcb89

Reply:
Well, with a better API this problem wouldn't even occur.

> this PR is getting old :P

7 days is nothing.^^

Reply:
I liked the idea of an iterator-based `write_bytes` enough that I made a PR for it: https://github.com/rust-lang/rust/pull/65621

Reply:
It looks really cool indeed :P

Reply:
Sadly [`Chain`](https://doc.rust-lang.org/std/iter/struct.Chain.html) does not implement `ExactSizeIterator` D:

Reply:
Also, writing with the new method:
```rust
       self.eval_context_mut()
           .memory
           .write_bytes(scalar, [bytes, b"0"].concat())
```
Has the same problem as before (the comment I deleted and rewrote, my bad):
```
error[E0080]: Miri evaluation error: attempted to read undefined bytes
   --> /home/christian/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libstd/ffi/c_str.rs:998:19
    |
998 |         let len = sys::strlen(ptr);
    |                   ^^^^^^^^^^^^^^^^ Miri evaluation error: attempted to read undefined bytes
    |
    = note: inside call to `std::ffi::CStr::from_ptr` at /home/christian/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libstd/sys/unix/os.rs:126:27
    = note: inside call to `std::sys::unix::os::getcwd` at /home/christian/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libstd/env.rs:50:5
note: inside call to `std::env::current_dir` at $DIR/current_dir.rs:8:15
   --> $DIR/current_dir.rs:8:15
    |
8   |     let cwd = env::current_dir().unwrap();
    |               ^^^^^^^^^^^^^^^^^^
    = note: inside call to `main` at /home/christian/.rustup/toolchains/miri/lib/rustlib/src/rust/src/libstd/rt.rs:61:34
```
I don't see the reason why this does not work and https://github.com/rust-lang/miri/commit/f7c6e0efbe11343562c713dce1d28e2d141dcb89 does.  On that commit I was already checking the access to memory anyway.

Reply:
> Sadly Chain does not implement ExactSizeIterator D:

Well that sucks.

@oli-obk any ideas? We could remove the `ExactSizeIter` bound from `write_bytes`; there are already assertions there checking that we are indeed writing the right number of bytes.

Reply:
@christianpoveda as for your other problem, just from a 3-line snippet I have no idea. Try debugging some more, in particular running Miri with tracing (`MIRI_LOG` env var). I won't be able to debug this until in about two weeks I am afraid.

Reply:
> Chain does not implement ExactSizeIterator D:

argml. this issue again :(

Well, we can add a custom chaining iterator that does implement `ExactSizeIterator`.

Alternatively we can just trust the `size_hint` and panic if it is wrong.

>  We could remove the ExactSizeIter bound from write_bytes; there are already assertions there checking that we are indeed writing the right number of bytes.

In that case, yes, let's remove the bound and just play a bit hardball with panics.

Reply:
> @christianpoveda as for your other problem, just from a 3-line snippet I have no idea. Try debugging some more, in particular running Miri with tracing (`MIRI_LOG` env var). I won't be able to debug this until in about two weeks I am afraid.

Here are the logs: https://gist.github.com/christianpoveda/b461680b73d2fbd266bfbace6c672a8e, I'll start digging into this for now.

Reply:
`b"0"` looks wrong... that's the digit 0 encoded as ASCII, not the `u8` 0!
This would explain the error you see as the code would go on reading memory until it finds a 0 terminator...

Reply:
Hahaha, we saw it almost at the same time. I'm correcting it now. What should I do about the `ExactSizeIterator` part?

Reply:
> What should I do about the ExactSizeIterator part?

For now you can keep using `concat` I guess.
https://github.com/rust-lang/rust/pull/65704 removes the bound.

Reply:
Ok then, do you want me to do the changes in this PR when https://github.com/rust-lang/rust/pull/65704 lands?